aten/src/ATen/native/RNN.cpp

```diff
@@ -0,0 +1,691 @@
+  result.reserve(vals.size() / 2);
+  for (int64_t i = 0; i < vals.size(); i += 2) {
+    result.emplace_back(vals[i], vals[i + 1]);
+  }
+  return result;
+}
+
+// Flattens a vector of pairs
+template<typename T>
+static std::vector<T> unpair_vec(std::vector<pair_of<T>>&& vals) {
+  std::vector<T> result;
+  result.reserve(vals.size() * 2);
+  for (int64_t i = 0; i < vals.size(); i++) {
+    result.push_back(std::move(vals[i].first));
+    result.push_back(std::move(vals[i].second));
+  }
+  return result;
+}
+
+// Parses a flat list of parameter tensors into a list of CellParams
+static std::vector<CellParams> gather_params(TensorList params, bool has_biases) {
+  static at::Tensor undefined;
+  std::vector<CellParams> result;
+  if (has_biases) {
+    AT_CHECK(params.size() % 4 == 0, "got an incorrect number of RNN parameters");
+    for (size_t i = 0; i < params.size(); i += 4) {
+      result.emplace_back(params[i], params[i + 1], params[i + 2], params[i + 3]);
+    }
+  } else {
+    AT_CHECK(params.size() % 2 == 0, "got an incorrect number of RNN parameters");
+    for (size_t i = 0; i < params.size(); i += 2) {
+      result.emplace_back(params[i], params[i + 1], undefined, undefined);
+    }
+  }
+  return result;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+// HIDDEN STATE FUNCTIONS
+//
+// Functions implemented below are implemented as templates based on hidden type,
+// because they need to work both with simple RNNs and GRU (which use a single Tensor),
+// as well as with LSTM (or possibly more complicated architectures in the future).
+// Still, there are some operations that need to be performed on the hidden states
+// alone, and for this purpose we provide an overloaded set of functions below.
+
+Tensor hidden_as_output(const Tensor& t) { return t; }
+Tensor hidden_as_output(const tpair_of<Tensor>& t) { return std::get<0>(t); }
+
+template<size_t index>
+std::vector<Tensor> project(at::ArrayRef<tpair_of<Tensor>> tuples) {
+  std::vector<Tensor> result;
+  result.reserve(tuples.size());
+  for (auto & t : tuples) {
+    result.push_back(std::get<index>(t));
+  }
+  return result;
+}
+
+Tensor hidden_concat(at::ArrayRef<Tensor> hiddens) { return at::cat(hiddens, 0); }
+tpair_of<Tensor> hidden_concat(at::ArrayRef<tpair_of<Tensor>> hiddens) {
+  return std::make_tuple(hidden_concat(project<0>(hiddens)), hidden_concat(project<1>(hiddens)));
+}
+
+Tensor hidden_slice(const Tensor& t, int64_t start, int64_t end) {
+  return t.narrow(0, start, end - start);
+}
+tpair_of<Tensor> hidden_slice(const tpair_of<Tensor>& t, int64_t start, int64_t end) {
+  return std::make_tuple(hidden_slice(std::get<0>(t), start, end),
+                         hidden_slice(std::get<1>(t), start, end));
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// CELL IMPLEMENTATIONS
+//
+// Cell is a basic component of an RNN, representing a single application of the
```

Comment:
nit: Move this comment below to Cell? You read it, and then you read `linear` and you're like, "This isn't a cell..."